flink-mesos/src/main/scala/org/apache/flink/mesos/scheduler/FlinkScheduler.scala

```diff
@@ -0,0 +1,223 @@
+    val requiredDisk = GlobalConfiguration.getFloat(
+      TASK_MANAGER_DISK_KEY, DEFAULT_TASK_MANGER_DISK)
+    val attributeConstraints = parseConstraintString(GlobalConfiguration.getString(
+      TASK_MANAGER_OFFER_ATTRIBUTES_KEY, DEFAULT_TASK_MANAGER_OFFER_ATTRIBUTES))
+    val role = GlobalConfiguration.getString(
+      MESOS_FRAMEWORK_ROLE_KEY, DEFAULT_MESOS_FRAMEWORK_ROLE)
+    val uberJarLocation = GlobalConfiguration.getString(
+      FLINK_UBERJAR_LOCATION_KEY, null)
+    val nativeLibPath = GlobalConfiguration.getString(
+      MESOS_NATIVE_JAVA_LIBRARY_KEY, DEFAULT_MESOS_NATIVE_JAVA_LIBRARY)
+
+    // combine offers into a single chunk
+    val totalMemory = offers.flatMap(_.getResourcesList
+      .filter(x => x.getName == "mem" && x.getRole == role)
+      .map(_.getScalar.getValue)).sum
+
+    val totalCPU = offers.flatMap(_.getResourcesList
+      .filter(x => x.getName == "cpus" && x.getRole == role)
+      .map(_.getScalar.getValue)).sum
+
+    val totalDisk = offers.flatMap(_.getResourcesList
+      .filter(x => x.getName == "disk" && x.getRole == role)
+      .map(_.getScalar.getValue)).sum
+
+    val portRanges = offers.flatMap(_.getResourcesList
+      .filter(x => x.getName == "ports" && x.getRole == role)
+      .flatMap(_.getRanges.getRangeList))
+
+    val ports = getNPortsFromPortRanges(2, portRanges)
+
+    val offerAttributes = toAttributeMap(offers.flatMap(_.getAttributesList))
+
+    // check if all constraints are satisfield
+    //  0. We need more task managers
+    //  1. Attribute constraints
+    //  2. Memory requirements
+    //  3. CPU requirements
+    //  4. Port requirements
+    val meetsRequirements =
+      taskManagers.size < maxTaskManagers &&
+      totalCPU >= requiredCPU &&
+      totalMemory >= requiredMem &&
+      totalDisk >= requiredDisk &&
+      ports.size == 2 &&
+      matchesAttributeRequirements(attributeConstraints, offerAttributes)
+
+    LOG.info( if(meetsRequirements) "Accepting" else "Declining " +
+      s"offer(s) from slave ${slaveId.getValue} " +
+      s"offered [cpus: $totalCPU | mem : $totalMemory | disk: $totalDisk] " +
+      s"required [cpus: $requiredCPU | mem: $requiredMem | disk: $requiredDisk]")
+
+    if (meetsRequirements) {
+      // create task Id
+      taskManagerCount += 1
+
+      // create executor
+      val command = createTaskManagerCommand(requiredMem.toInt)
+      val log4jUrl = s"${httpConfigServerAddress.get}/log4j.properties"
+      val executorInfo = createExecutorInfo(s"$taskManagerCount", role,
+        Set(uberJarLocation, log4jUrl), command, nativeLibPath)
+
+      // create task
+      val taskId = TaskID.newBuilder().setValue(s"TaskManager_$taskManagerCount").build()
+      val taskInfo = createTaskInfo(
+        "taskManager", taskId, slaveId, role, requiredMem,
+        requiredCPU, requiredDisk, ports, executorInfo, currentConfiguration.get)
+
+      Seq(taskInfo)
+    } else {
+      Seq()
```

Comment:
There has a bug, if you not call driver.declineOffer(offer.id) mesos will hold this offer for FlinkScheduler, then other framework could not use this offer and there is no resource any more.